Formstack Baton requests error handling

ob6160 opened the pull request:
## What does this change?
This PR makes the process make sure the submissions referenced in dynamo still exist in the formstack account they are expected to be.
- If a submission is not found in account 2 we look for it in account 1 before giving up
- every time we fetch a submission from formstack we double check the submission belongs to the email in the original request

pvighi commented on formstack-baton-requests/src/main/scala/com/gu/identity/formstackbatonrequests/services/FormstackService.scala:
```diff
@@ -79,24 +79,33 @@ object FormstackService extends FormstackRequestService with LazyLogging {
+
+      // Define errors that we can safely skip
+      def isDecryptionError = response.body.contains("An error occurred while decrypting the submissions")
+      def isSubmissionDeletedError = response.body.contains("An error occurred while decrypting the submissions")
```
this is probably the wrong error message for this, right ?

kenoir replied:
Docs seem to [indicate that it'll 404 in this case](https://developers.formstack.com/reference/submission-id-get), and the error will look like:

```json
{
  "status": "error",
  "error": "A valid submission id was not supplied"
}
```

pvighi commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Make baton-formstack handle form deletions](https://trello.com/c/ZQmesRDb/3890-make-baton-formstack-handle-form-deletions)

pvighi commented on formstack-baton-requests/src/main/scala/com/gu/identity/formstackbatonrequests/services/FormstackService.scala:
```diff
@@ -79,24 +79,33 @@ object FormstackService extends FormstackRequestService with LazyLogging {
-        Left(FormstackDecryptionError(s"${response.body} | submission id: ${submissionIdEmail.submissionId}"))
-      else
-        decode[Submission](response.body)
+      def isSkippableError = isErrorResponse && isDecryptionError && isSubmissionDeletedError && isFormNotFoundError
```
It should be one of the skippable errors, not all at the same time
```suggestion
      def isSkippableError = isErrorResponse && (isDecryptionError || isSubmissionDeletedError || isFormNotFoundError)
```

pvighi committed:
added skippable errors that don't cause the whole process to fail

pvighi committed:
added skippable errors that don't cause the whole process to fail

pvighi committed:
added skippable errors that don't cause the whole process to fail

pvighi commented on formstack-baton-requests/src/main/scala/com/gu/identity/formstackbatonrequests/services/FormstackService.scala:
```diff
@@ -137,7 +146,7 @@ object FormstackService extends FormstackRequestService with LazyLogging {
   override def deleteUserData(submissionIdEmails: List[SubmissionIdEmail], config: PerformLambdaConfig): Either[Throwable, List[SubmissionDeletionReponse]] = {
```
I can't really think of  a way to test deleting data as it seems we only have a PROD environment. I tried to share as much code with `getSubmissions` to at least have some testing on it

ob6160 replied:
I can't either, the best suggestion I've heard so far is to set up a test account in Formstack and run tests against that locally perhaps

pvighi commented on formstack-baton-requests/src/main/scala/com/gu/identity/formstackbatonrequests/services/FormstackService.scala:
```diff
@@ -74,30 +75,38 @@ object FormstackService extends FormstackRequestService with LazyLogging {
-
+  
+  val skippableErrorMessages = List(
+    "An error occurred while decrypting the submissions",
```
@ob6160  I don't remember why this was added ( I think it was a previous issue) but are we sure we want to skip this error ?

kenoir replied:
No, we're not. It's not clear in what situations this error would arise.

kenoir replied:
For this PR though we could keep it the same, to avoid changing current behaviour.

pvighi replied:
I think the current behaviour might be wrong..
It catches the "decoding" error but then it still returns a `Left` with a custom error that doesn't get any special treatment later so it still fails the whole thing, just with a different error:
https://github.com/guardian/identity-processes/pull/90/files

pvighi replied:
I think it was modelled after the behaviour when dynamodb is updated with new form data but I think that case it works because:

1. The function that returns the `Left` only makes a single api call so it's fine to return a `Left`. In this one we are in a loop so we want to continue with the next submission and  return a `Right` with a list of all the successful submissions unless there's an unskippable error in which case we would want to return a `Left`
2. The function that calls that has a special case to deal with the custom `FormstackDecryptionError` and not return an error. See [Here](https://github.com/guardian/identity-processes/blob/e92e3ae5becc2419769d35159cb3924c8a23afaa/formstack-baton-requests/src/main/scala/com/gu/identity/formstackbatonrequests/services/DynamoUpdateService.scala#L39)

ob6160 replied:
i think we can remove it from the list because on line 73 we have the check from before we came to the codebase: https://github.com/guardian/identity-processes/pull/92/files#diff-2e97f551ddedcd9a50db834124cfb109ad0dc4f748e72de9f9b84623d8625f0bR73

pvighi commented on formstack-baton-requests/src/local/com/gu/identity/formstackbatonrequests/services/FormstackRequestServiceLocalRun.scala:
```diff
@@ -0,0 +1,36 @@
+  println("response was:")
+  println(response)
+
+}
```
I added this quick test that calls the part of the code that I want to modify. Unfortunately we can only do this for the  function used to process SARs, the RER one has a very similar change but I cannot run this without actually deleting something from PROD.

pvighi commented on formstack-baton-requests/src/main/scala/com/gu/identity/formstackbatonrequests/services/FormstackService.scala:
```diff
@@ -74,30 +75,38 @@ object FormstackService extends FormstackRequestService with LazyLogging {
-
+  
+  val skippableErrorMessages = List(
+    "An error occurred while decrypting the submissions",
```
```suggestion
```
In short I think we should probably not skip this error. What do you think @ob6160 ?

ob6160 replied:
okay, let's not skip it - if it turns out that it breaks things it's not a big deal to add it back in

pvighi commented on formstack-baton-requests/src/main/scala/com/gu/identity/formstackbatonrequests/services/FormstackService.scala:
```diff
@@ -74,30 +75,41 @@ object FormstackService extends FormstackRequestService with LazyLogging {
-      /* There are a couple of forms that the Formstack API can't seem to decrypt. There seems to be no way around this
-      *  so we capture this specific error and skip these forms. */
-      if(response.body.contains("An error occurred while decrypting the submission"))
-        Left(FormstackDecryptionError(s"${response.body} | submission id: ${submissionIdEmail.submissionId}"))
```
just to sum up our previous conversation for future reference: 

Going from the comment in the code  it seems like the intention was to skip this error and keep going, but the current code wasn't doing doing that, it was just returning a custom error type that nothing picked up later.
We don't think this error should be skipped anyway, but if we later decide it should be we can add it to the list of skippable errors

Commit:
curry writeSubmissionsPage

Commit:
Add happy path test for updateSubmissionsTable

Commit:
add test for execution timeout

Commit:
remove unnecessary wrapping statement

Commit:
adds tests for failure conditions

Commit:
little reformat

Commit:
Refactor writeSubmissionsPage to be easier to read

Commit:
remove recursion from writing pages

Commit:
simplify

Commit:
Parallelise writing submission pages

ob6160 committed:
fix: add skip for password protected forms

Commit:
type errors and add test

Commit:
remove unnecessary import

NovemberTang committed:
Add snyk monitoring for main branch

dblatcher committed:
change the test to use another consent email

dblatcher committed:
rename Holidays to Traveller and make it a 'set-list' for the identy name of the gaurdian traveller editorial newsletter

dblatcher committed:
update readme with a not about usage

pvighi committed:
wip unit test for formstackservice

The http dependency had to be extracted so that we could replace it with a mock for tests.

pvighi committed:
wip unit test for formstackservice

The http dependency had to be extracted so that we could replace it with a mock for tests.

pvighi committed:
fallback mechanism and email validation

if a submission is not found in account 1 try fetching it from account 2. Also validate that the submissions coming from Formstack actually contain references to the email we are looking for, otherwise ignore them

pvighi committed:
fallback mechanism and email validation

if a submission is not found in account 1 try fetching it from account 2. Also validate that the submissions coming from Formstack actually contain references to the email we are looking for, otherwise ignore them

pvighi committed:
fix account number in RER processing

pvighi committed:
fix account number in RER processing

pvighi committed:
Merge branch 'ob-pv/skip-formstack-submissions-on-error' into pv/formstack_baton_error_handling

pvighi committed:
made the email verification case insensitive and added more tests

pvighi committed:
Merge branch 'pv/formstack_baton_error_handling' of github.com:guardian/identity-processes into pv/formstack_baton_error_handling

pvighi committed:
fix a bug with the hashed emails

also some refactoring of the case classes in formstackservice

pvighi committed:
made the formstack service a regular class instead of a case class

pvighi committed:
fixed tests

pvighi committed:
Merge pull request #102 from guardian/pv/formstack_baton_error_handling

Pv/formstack baton error handling

pvighi committed:
undo changes in delete process we cannot really test

we are feeding it validated now data so we don't need an extra check

pvighi commented:
![](https://github.trello.services/images/mini-trello-icon.png) [formstack migration baton changes](https://trello.com/c/r6zXHbjp/4180-formstack-migration-baton-changes)

pvighi commented on formstack-baton-requests/src/main/scala/com/gu/identity/formstackbatonrequests/services/FormstackService.scala:
```diff
@@ -129,25 +170,74 @@ object FormstackService extends FormstackRequestService with LazyLogging {
+
+  }
+
+  def deleteValidatedSubmissions(submissionIdEmails: List[SubmissionIdEmail], config: PerformLambdaConfig): Either[Throwable, List[SubmissionDeletionReponse]] = {
```
the idea of this is that since we cannot test deletions we are just keeping the existing deletion code intact and putting some validation phase before that we can test. Now this should recieve a list of submissions that are supposed to be "fixed": submissions that refer to forms that don't exist are removed,  if dynamo has the wrong account id for a submission it is fixed and if any submission doesn't contain a reference to the original email in the request it is also removed

kelvin-chappell reviewed (approved):
It's difficult to follow the whole thing but the latest changes look fine to me 👍 
If we have to maintain this code in the long term we'll need to make it, and the environment, a bit easier to work with.

pvighi committed:
renamed some vals for clarity

pvighi committed:
added extra test
